kubos/utils/sdk.py

```diff
@@ -13,18 +13,106 @@
     if attr in sdk_data:
         return sdk_data[attr]
     else:
         return None
+
+def get_module_name(path):
+    if not os.path.isfile(path):
+        return None
+    string = open(path, 'r').read()
+    data = json.loads(string)
+    return data['name']
+
+def is_marker(entity_name):
+    #Determine if the directory name is at the root of a module or target
+    marker_names = ['module.json',
+                    'target.json']
+    return True if entity_name in marker_names else False
+
+
+'''
+The following functions are used for recursively creating links too and from the global
+cache (located at /usr/local/lib/yotta_*).
+
+`kubos update` command will link all modules
+to the global cache after downloading newer kubos source.
+
+`kubos link --all` will link all of the modules and target in the global cache into a
+project.
+'''
+
+
+def link_entities(src, dst):
+    '''
+    This function is used for recursively searching subdirectories for module and target
+    roots ie. where a module.json or target.json file exists once a root is found, the 
+    linking function is called.
+    '''
+    logging.disable(logging.WARNING) #suppress yotta warning for linking non-required modules and targets
+    for subdir in os.listdir(src):
+        #loop through the subdirectories of src
+        cur_dir = os.path.join(src, subdir)
+        if is_marker(subdir):
+            #if we're pointing to a target.json or module.json - link the module and return
+            #NOTE: This assumes there are not nested modules
+            run_link(cur_dir, dst)
+            return
+        elif os.path.isdir(cur_dir):
+            #if we're looking at a subdirectory recursively search for a module root
+            link_entities(cur_dir, dst)
+
+
+def run_link(src, dst):
+    '''
+    This is used for proxying to the yotta link and link_target commands that actually
+    generate the links to or from the global cache.
+    '''
+    link_module = link if 'module' in src else link_target
+    if dst:
+        #we're linking to a project from the global cache so we need to link the module by name
+        entity_name = get_module_name(src)
+        path = dst
+    else:
+        #we're linking to the global cache, the default behavior of a None module/target name
+        #is linking it to the global cache.
+        entity_name = None
+        path = os.path.dirname(src)
+    start_dir = os.getcwd()
+    os.chdir(path)
+    link_args = argparse.Namespace(module_or_path=entity_name,
+                                   target_or_path=entity_name,
+                                   config=None,
+                                   target=yotta.lib.detect.systemDefaultTarget(),
+                                   save_global=True,
+                                   no_install=False)
+    link_module.execCommand(link_args, '')
+    os.chdir(start_dir)
+
+
+def link_to_project(project):
+    print 'Linking modules...'
```

Comment:
I would suggest we use the `logging` module and route all our output through a `logger.level()` call. This will give us a bit more control over the debug vs normal output.